Add an elasticsearch span store.

anuraaga opened the pull request:
Also fixes a typo in SpanStoreTest that made it less stringent than the scala version.
Also fixes DependencyTest to use non-ordered assertions.
Also adds single DependencyLink serialization methods to JsonCodec.

Closes #103

adriancole commented on zipkin-server/pom.xml:
```diff
@@ -179,6 +179,7 @@
         <artifactId>spring-boot-maven-plugin</artifactId>
         <configuration>
           <classifier>exec</classifier>
+          <layout>ZIP</layout>
```
why is this needed?

anuraaga replied:
It switches spring-boot's loader to PropertiesLoader, which allows adding things to the classpath if you want. By making the spanstores optional in zipkin-server, this seems like the best way to allow people to run with their favorite spanstore without having to build everything from source.

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchProperties.java:
```diff
@@ -0,0 +1,79 @@
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties("elasticsearch")
```
span store shouldn't have spring deps

adriancole replied:
well, right now we don't anyway.. I suppose we could host things like spring config as optional classes in the jar, just that it makes for a leaky classpath when coding and writing tests. regardless, I think re-organizing how modules are loaded into zipkin-server would be better a separate issue

adriancole commented on zipkin-spanstores/elasticsearch/pom.xml:
```diff
@@ -0,0 +1,117 @@
+    </profile>
+  </profiles>
+
+  <dependencies>
```
please roll this back to a simple classpath. span stores are meant to be used in any framework.

anuraaga replied:
Do you mean not using spring-boot to package these? What if I add a classifier like the server binary? Otherwise I'm not sure of a good way to actually start the server with this and all its dependencies in the classpath.

adriancole replied:
As a first step, just mimicking Cassandra is probably best.

Make the dep optional in zipkin-server (we have an integration test to show
it works without). Make a buddy configuration like the one for Cassandra.

In a separate issue, we can work on a different way to manage the
classpath, since it applies to all span stores and transports.

adriancole replied:
PS the Cassandra optional thing is new as of today! Pull latest

adriancole commented:
Pretty impressed you've gotten this far with no help! Nice. Thanks for putting this together.

I made some comments, basically that we should separate the concerns about how to wire the server from adding elasticsearch support. Can you put it back like the others first and then open a separate issue on modularity? I don't like merging partial work (ex make this work but not another span store), removing this concern makes it a simpler change to review, and more easy to see when we try for the other span stores.

(ex. spring config could be a "buddy" module or something, but I really don't want the span stores to have a compile dep, as that makes it hard even to upgrade spring)

adriancole commented:
PS nevermind the scala tests that don't work (right now anyway). I've a
TODO (in my mind) to port the remaining to java.

anuraaga commented:
Ah I get your intention with the span stores (I was just treating them as optional parts of zipkin-server rather than something more broadly usable). Will move to following the current pattern.

adriancole commented:
ps jump on https://gitter.im/openzipkin/zipkin any time. I'm sure your efforts will prove quite popular! cc @shakuzen @trustin

adriancole commented on zipkin-server/pom.xml:
```diff
@@ -91,6 +91,20 @@
+      <optional>true</optional>
+    </dependency>
+
+    <dependency>
```
this should be transitive from spanstore-elasticsearch, right?

anuraaga replied:
Intellij seems to not want to go transitive on this. Do you know if there's a better way (my maven-fu isn't up to speed)?

For now I have added a comment that this is for intellij (don't know if eclipse is affected)

adriancole replied:
no problem.. comments are fine.. any of us can dig into it later that way.

adriancole commented on interop/pom.xml:
```diff
@@ -74,10 +74,24 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
```
I don't think this is used explicitly in the code, so probably better to keep this out. If this is needed because of some dependency clash, add a comment what's causing that.

adriancole commented on interop/src/test/java/zipkin/elasticsearch/ElasticsearchTestGraph.java:
```diff
@@ -0,0 +1,44 @@
+    if (this.spanStore == null) {
+      try {
+        this.spanStore = new ElasticsearchSpanStore(CONFIG);
+      } catch (NoHostAvailableException e) {
```
c&p error. If the constructor doesn't do I/O, you won't need this.

adriancole commented on zipkin-server/src/main/java/zipkin/server/ZipkinElasticsearchProperties.java:
```diff
@@ -0,0 +1,79 @@
+   * The date format string to use to generate daily index names, as used with Java
+   * DateTimeFormatter. Defaults to "yyyy-MM-dd".
+   */
+  private String dailyIndexFormat = "yyyy-MM-dd";
```
seems quite a special-case configuration to expose. unless we think this is important enough to test and support incompatibilities when people change it, I'd punt.

adriancole commented on zipkin-server/src/main/java/zipkin/server/ZipkinServerConfiguration.java:
```diff
@@ -71,6 +73,12 @@ Sampler traceIdSampler(@Value("${zipkin.collector.sample-rate:1.0}") float rate)
   @ConditionalOnMissingBean(SpanStore.class)
   SpanStore spanStore() {
+    if (server.getStore().getType() != ZipkinServerProperties.Store.Type.mem) {
+      throw new IllegalStateException("Attempted to set storage type to "
```
agree with the exception, just not the advice as yet. Ex. you could mention there's a potential for classpath issue but the fix isn't yet adding one jar to LOADER_PATH. (we could make the latter the answer in a separate change)

adriancole replied:
looks fixed

adriancole commented on zipkin-server/src/main/java/zipkin/server/ZipkinServerConfiguration.java:
```diff
@@ -132,6 +140,22 @@ public Object postProcessAfterInitialization(Object bean, String beanName) {
 
+  @Configuration
+  @EnableConfigurationProperties(ZipkinElasticsearchProperties.class)
+  @ConditionalOnProperty(name = "zipkin.store.type", havingValue = "elasticsearch")
```
nice!

adriancole commented on zipkin-server/src/main/resources/zipkin-server.yml:
```diff
@@ -23,6 +23,12 @@ mysql:
+  cluster-name: ${ES_CLUSTER_NAME:elasticsearch}
+  hosts: ${ES_HOSTS:localhost:9300}
+  index: ${ES_INDEX:zipkin}
+  document-type: ${ES_DOCUMENT_TYPE:span}
```
I'd go with minimum settings for now.. let people ask for things like document-type or daily-index-format. it will reduce the amount of tests we have to maintain

anuraaga replied:
Had actually already removed documentType with the same idea, removed the index format as well. I guess there are multiple competing formats in the elasticsearch world (logstash and the rest IIUC), but it probably doesn't really matter for zipkin.

adriancole commented on zipkin-spanstores/elasticsearch/README.md:
```diff
@@ -0,0 +1,19 @@
+This is an Elasticsearch 2 SpanStore. The SpanStore utilizies the Elasticsearch Java client
+library with a node client for optimal performance.
+
+This spanstore is not included in the zipkin-java binary by default. To use it, add it to the
```
comment made already, but let's keep these instructions to the side, and let's resolve them for all span stores.

adriancole commented on zipkin-spanstores/elasticsearch/pom.xml:
```diff
@@ -0,0 +1,85 @@
+  <parent>
+    <groupId>io.zipkin.java</groupId>
+    <artifactId>zipkin-spanstores</artifactId>
+    <version>0.9.1-SNAPSHOT</version>
```
hehe 0.9.2-SNAPSHOT now :D

adriancole commented on zipkin-spanstores/elasticsearch/pom.xml:
```diff
@@ -0,0 +1,85 @@
+
+    <dependency>
+      <groupId>org.elasticsearch.plugin</groupId>
+      <artifactId>license</artifactId>
```
what does this do?

anuraaga replied:
Added a comment, it allows the client to connect to clusters that are using licensed plugins (like marvel, watcher, etc).

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanConsumer.java:
```diff
@@ -0,0 +1,90 @@
+      request.add(createSpanIndexRequest(ApplyTimestampAndDuration.apply(span)));
+    }
+    ListenableActionFuture<BulkResponse> future = request.execute();
+    if (BLOCK_ON_FUTURES) {
```
cool I'm glad this is possible, as making tests deal with eventual consistency is a pita

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanConsumer.java:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  private IndexRequestBuilder createSpanIndexRequest(Span span) {
+    Instant indexTimestamp;
```
this seems pretty fancy if the goal is to get the format of the day of the timestamp :P

adriancole replied:
main thing I'm thinking about is that while we've not enforced it, the code will have higher re-use if it doesn't use java 8 apis. This particular task isn't too hard even with normal java Date.

anuraaga replied:
Ah I noticed zipkin-server is java8 and assumed the spanstores were too (was still thinking they're just components). Do you think I should go ahead and target java7?

adriancole replied:
>
> Ah I noticed zipkin-server is java8 and assumed the spanstores were too (was still thinking they're just components). Do you think I should go ahead and target java7?
>
> I would go for 7 at the moment. I can follow-up with the others as well.
Apologies for the sloppiness on things like this :P

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,375 @@
+  }
+
+  @Override public List<List<Span>> getTraces(QueryRequest request) {
+    Instant end = Instant.ofEpochMilli(request.endTs);
```
I'd prefer not dancing in and out of Instants. I might be missing something, but it looks like we're just doing subtraction, and the '-' operator is more straightforward.

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,375 @@
+    for (Map.Entry<String, String> annotation : request.binaryAnnotations.entrySet()) {
+      filter.must(nestedQuery("binaryAnnotations",
+          boolQuery()
+              .must(termQuery("binaryAnnotations.key", annotation.getKey()))
```
check type = string

anuraaga replied:
We actually don't store type for strings in the default JsonCodec. I was very happy to be able to achieve this without writing my own json serialization for elasticsearch :) Making this explicit would require custom serialization, not sure if it's worth it. Implicitly, this can only match a string value since in the index template, we set the type of the field to string and ignore non-string fields when indexing, so a non-string value cannot be matched by nature of the index.

Would a comment on this line help?

adriancole replied:
yep. just comment it!

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,375 @@
+            .stream()
+            .map(id -> String.format("%016x", id))
+            .collect(Collectors.toList())))
+        // We sort descending here so traces are sorted in order of descending timestamp while we
```
notably this makes limit apply correctly

anuraaga replied:
Nah that's handled by the size limit on the aggregation itself around line 150. I've added a comment there to make it a bit clearer.

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,375 @@
+            .setTypes(ElasticsearchConstants.SPAN)
+            .setQuery(matchAllQuery())
+            .setSize(0)
+            .addAggregation(AggregationBuilders.terms("annotationServiceName_agg")
```
naming conventions or other schema like things might be nice to document in the readme or at the top of this file.

adriancole replied:
if this term name is transient for the query, ignore me.

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,375 @@
+              try {
+                return JsonCodec.DEPENDENCY_LINK_ADAPTER.fromJson(hit.getSourceAsString());
+              } catch (IOException e) {
+                throw JsonCodec.exceptionReading(
```
unless there's a test that doing this helps not trip, I'd not yank this internal function :) just throw your own?

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,375 @@
+    try {
+      Class.forName(LicensePlugin.class.getName());
+    } catch (ClassNotFoundException ignored) {
+      throw new IllegalStateException("Can't happen");
```
haha you wait!

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/IndexNameFormatter.java:
```diff
@@ -0,0 +1,32 @@
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+
+final class IndexNameFormatter {
```
made a note before, but I'd prefer to lock in a format and document it. Most span stores don't offer this flexibility and it will be easier to support with only very high-level knobs.

anuraaga replied:
This class was mainly to avoid typing index + "-" + format(timestamp) all over the place. I realized I could clean things up a bit more which also makes more obvious the date format isn't configurable anymore just a constant defined here.

adriancole commented:
ok finished a wave of review! again quite impressed you've managed to progress so quickly

anuraaga commented:
Thanks for the reviews, addressed most comments will spend some time on java7ifying (including removing Instant assuming I can do the day truncation math properly ;) ).

adriancole commented:
I think there's a day truncation fn used in Cassandra span store (floor to
day or something).

Will take another look tomorrow. Keep it up!

adriancole commented:
Util.midnightUTC is the fn I was thinking about.

adriancole commented:
opened https://github.com/openzipkin/zipkin-java/issues/108 to address the modularity as a separate effort (so that we can apply to all span stores)

anuraaga commented:
Targets java7 now, added travis settings, and all tests including scala pass now! I found that the Java equivalent of getTraces_duration had an incorrect assertion so fixing that allowed it to fail as it was supposed to which made it much easier to debug :)

I realized my trace ordering logic was fairly wrong, as it was ordering on timestamp of first span that actually passed the filter, even though we want to order on first span of the trace, period. I was worried this would require an additional search to deal with but after purusing the elasticsearch docs I was able to come up with something. I'm not sure if it's elegant or maniacal, but at least it's just one search...

adriancole commented:
Sorry about the confusing bad assertion!

I'm so happy to hear you have all the things passing. I will take a look
tomorrow morning or sooner.

adriancole commented on interop/src/test/java/zipkin/elasticsearch/ElasticsearchTestGraph.java:
```diff
@@ -0,0 +1,39 @@
+    ElasticsearchSpanConsumer.BLOCK_ON_FUTURES = true;
+  }
+
+  private AssumptionViolatedException ex;
```
this is private, and I don't see this being set.. remove?

adriancole commented on interop/src/test/java/zipkin/elasticsearch/ElasticsearchTestGraph.java:
```diff
@@ -0,0 +1,39 @@
+  private AssumptionViolatedException ex;
+  private ElasticsearchSpanStore spanStore;
+
+  /** A lot of tech debt here because the repository constructor performs I/O. */
```
I don't see the tech-debt here :) c&p?

anuraaga replied:
I'm actually not sure exactly what this comment is talking about. ElasticsearchSpanStore constructor does perform I/O - it connects to the cluster and also checks for the presence of the important zipkin index template. So I guess I should leave this as is?

adriancole replied:
Sure. I thought I saw an assumption violated exception not being set.
Basically, we need to capture the error when the cluster isn't available
and convert that to an assumption violated exception. Maybe kill your local
ES process.

anuraaga replied:
Ah I get it now, thanks! Let me add back AssumptionViolatedException which is thrown on NoNodeAvailableException as that's the equivalent.

I was always wondering how these cassandra tests were passing on my machine, I wasn't familiar with this mechanism - cool.

anuraaga replied:
Fixed to convert to AssumptionViolatedException as it should. Now tests are skipped when elasticsearch isn't running locally.

adriancole commented on interop/src/test/java/zipkin/elasticsearch/ElasticsearchTestGraph.java:
```diff
@@ -0,0 +1,39 @@
+  synchronized ElasticsearchSpanStore spanStore() {
+    if (ex != null) throw ex;
+    if (this.spanStore == null) {
+      this.spanStore = new ElasticsearchSpanStore(CONFIG);
```
if this doesn't do i/o, you can just set it as a final reference. Enum itself will handle the only-once synchonization.

adriancole commented on zipkin-server/src/main/java/zipkin/server/ZipkinServerConfiguration.java:
```diff
@@ -43,13 +43,15 @@
 import zipkin.server.brave.TraceWritesSpanStore;
 
 @Configuration
-@EnableConfigurationProperties({ZipkinServerProperties.class, ZipkinCassandraProperties.class})
```
thx

adriancole commented on zipkin-server/src/main/resources/zipkin-server.yml:
```diff
@@ -23,6 +23,11 @@ mysql:
+  cluster-name: ${ES_CLUSTER_NAME:elasticsearch}
+  hosts: ${ES_HOSTS:localhost:9300}
+  index: ${ES_INDEX:zipkin}
+  daily-index-foramt: ${ES_DAILY_INDEX_FORMAT:yyyy-MM-dd}
```
rm .. old ref

adriancole commented on zipkin-spanstores/elasticsearch/README.md:
```diff
@@ -0,0 +1,13 @@
+library with a node client for optimal performance.
+
+Spans are stored into daily indices, for example spans with a timestamp falling on 2016/03/19
+will be stored in an index like zipkin-2016-03-19. There is no support for TTL through this SpanStore.
```
nice

adriancole commented on zipkin-spanstores/elasticsearch/pom.xml:
```diff
@@ -0,0 +1,113 @@
+      <version>${elasticsearch.version}</version>
+    </dependency>
+
+    <!-- For scripting. -->
```
ick. groovy is a pretty large dependency. maybe be more specific about the functionality that needs this?

anuraaga replied:
Was able to get away with a lucene expression, so gone!

adriancole replied:
:clap:

adriancole commented on zipkin-spanstores/elasticsearch/pom.xml:
```diff
@@ -0,0 +1,113 @@
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <version>1.7.19</version>
```
extract slf4j.version property por favor

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,470 @@
+    // aggregation which applies the filter, and then removes parent buckets based on whether any
+    // documents matched the filter, effectively "backtracking".
+
+    // Pipeline aggregations don't have a Java API yet, so build it in json.
```
is there an issue to make a java api for this? if so, link it?

anuraaga replied:
Found the Java API phew

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,470 @@
+    // documents matched the filter, effectively "backtracking".
+
+    // Pipeline aggregations don't have a Java API yet, so build it in json.
+    XContentBuilder filterAggJson = null;
```
extract this into a method since it has a different api style than other parts here.. makes it a bit less distracting to read that way.

anuraaga replied:
It's using Java API now and is much simpler so left inline.

adriancole commented on zipkin-spanstores/elasticsearch/src/test/java/zipkin/elasticsearch/ElasticsearchTestGraph.java:
```diff
@@ -0,0 +1,33 @@
+enum ElasticsearchTestGraph {
+  INSTANCE;
+
+  static final ElasticsearchConfig CONFIG = new ElasticsearchConfig.Builder().build();
```
enum is implicitly static singleton, see if you can get away with just sharing a final reference.

```java
final ElasticsearchSpanStore spanStore  new ElasticsearchSpanStore(new ElasticsearchConfig.Builder().build());
```

adriancole commented:
few things left I think..

* document the new env variables in zipkin-server/README (and manually test at least host picks up)
* check into groovy.. see if this is something that is not avoidable. one of the java repo's goals was to provide a dependency light alternative to the scala one.
* otherwise, some minor cleanups.

Thanks again for the hard work!

adriancole commented:
Added this to the 1.0 milestone, not because elasticsearch is in zipkin-scala, as it isn't. Not even because this was asked for many times over the last several months. I'm adding this to 1.0 because it tests our apis and modularity.

adriancole commented on zipkin-spanstores/elasticsearch/pom.xml:
```diff
@@ -0,0 +1,113 @@
+  <parent>
+    <groupId>io.zipkin.java</groupId>
+    <artifactId>zipkin-spanstores</artifactId>
+    <version>0.9.2-SNAPSHOT</version>
```
sorry, but we had a nasty UI bug. this needs to bump to 0.9.3-SNAPSHOT next time you rebase

anuraaga replied:
Thanks for the heads up! Will wrap this up tonight or tomorrow, have an
idea for groovy that I hope works (relies on the fact that ES docs appear
to contradict each other, wonder which one is right ;) )
On Mar 21, 2016 12:10 PM, "Adrian Cole" <notifications@github.com> wrote:

> In zipkin-spanstores/elasticsearch/pom.xml
> <https://github.com/openzipkin/zipkin-java/pull/107#discussion_r56779554>:
>
> > +
> > +    http://www.apache.org/licenses/LICENSE-2.0
> > +
> > +    Unless required by applicable law or agreed to in writing, software distributed under the License
> > +    is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
> > +    or implied. See the License for the specific language governing permissions and limitations under
> > +    the License.
> > +
> > +-->
> > +<project xmlns="http://maven.apache.org/POM/4.0.0"
> > +         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
> > +         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
> > +  <parent>
> > +    <groupId>io.zipkin.java</groupId>
> > +    <artifactId>zipkin-spanstores</artifactId>
> > +    <version>0.9.2-SNAPSHOT</version>
>
> sorry, but we had a nasty UI bug. this needs to bump to 0.9.3-SNAPSHOT
> next time you rebase
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/openzipkin/zipkin-java/pull/107/files/e29df6931a9583f93e6a3331e7788962350870c0#r56779554>
>

adriancole replied:
ok one more bump :) 0.9.4-SNAPSHOT
things to note on rebase
* travis is on a larger slave, now. have a second look, but hopefully it can get test execution reliable
* all tests have been ported from scala

anuraaga commented:
Addressed comments, only one left that I'm not clear on is related to cleaning up ElasticsearchTestGraph or not.

Also, thanks for preempting me as I was indeed worried I had hit a travis resource limit :)

anuraaga commented:
Thanks for explaining the test graph portion - now I think the PR is ready for another round

adriancole commented:
ok I'm going to try to run it, this time.

adriancole commented on zipkin-server/README.md:
```diff
@@ -58,6 +58,24 @@ Example usage:
+### Elasticsearch
+The following apply when `STORAGE_TYPE` is set to `elasticsearch`:
+
+    * `ES_CLUSTER_NAME`: The name of the elasticsearch cluster to connect to. Defaults to "elasticsearch".
```
nit: unless _NAME disambiguates, I'd drop it

adriancole commented on zipkin-spanstores/elasticsearch/pom.xml:
```diff
@@ -0,0 +1,99 @@
+          </snapshots>
+          <id>elasticsearch</id>
+          <name>releases</name>
+          <url>http://maven.elasticsearch.org/public-releases/</url>
```
are these not synced to maven central? if they are, I'd prefer to remove this profile, as it will distract people.

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchConfig.java:
```diff
@@ -0,0 +1,65 @@
+ */
+package zipkin.elasticsearch;
+
+import static java.util.Objects.requireNonNull;
```
animal sniffer, if enabled, would catch this :) Use Util.checkNotNull

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,447 @@
+      try {
+        client.addTransportAddress(new InetSocketTransportAddress(InetAddress.getByName(
+            hostAndPort.getHostText()), hostAndPort.getPort()));
+      } catch (UnknownHostException e) {
```
careful.. I might ask you to test something like this :)
no action needed now, but I'm only half-joking. Let's think about a test that can verify things like this.

adriancole commented on interop/src/test/java/zipkin/elasticsearch/ElasticsearchScalaDependencyStoreTest.java:
```diff
@@ -0,0 +1,57 @@
+import com.twitter.zipkin.common.Span;
+import com.twitter.zipkin.storage.DependencyStore;
+import com.twitter.zipkin.storage.DependencyStoreSpec;
+import java.time.Instant;
```
check for unused imports. I think there are a few files that import this

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,447 @@
+    ImmutableList.Builder<Span> trace = ImmutableList.builder();
+    for (SearchHit hit : response.getHits()) {
+      try {
+        trace.add(JsonCodec.SPAN_ADAPTER.fromJson(hit.getSourceAsString()));
```
if you use Codec.JSON.readSpan() you wouldn't need the try catch.

adriancole replied:
unless getSourceAsString throws..

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,447 @@
+    ArrayListMultimap<Long, Span> groupedSpans = ArrayListMultimap.create();
+    for (SearchHit hit : response.getHits()) {
+      Span span;
+      try {
```
same here.. I'd rather overload `JsonCodec.readSpan(String)` if you are looking for a means to avoid turning a string to a byte array. We did something similar with `ThriftCodec.readSpan(ByteBuffer)`

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,447 @@
+      // We would have no bucket if there wasn't a hit, so this should always be non-empty.
+      SearchHit hit = hitsAgg.getHits().getAt(0);
+      DependencyLink link;
+      try {
```
same comment as SPAN_ADAPTER

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,447 @@
+    if (!existingTemplates.getIndexTemplates().isEmpty()) {
+      return;
+    }
+    String template;
```
this should be loaded once

adriancole commented:
Cool .. Nice work! Less deps and passing travis! two main comments I think
we could address before merge.

First, modify json codec, or add a TODO vs using the moshi adapters
directly. I think these aren't really intended to be exposed, but are to
make toString easy elsewhere.

Second: move the static configuration to the config class even if we don't
expose a builder.

Ex the index formats and template (loaded from classpath) could be getters
on the config object. This should reduce some clutter and the need for
extra files.

There may be other refactoring opportunities later, but we can address
those post merge.

anuraaga commented:
Fixed the json adapter stuff (found how to get the bytes directly from the elasticsearch result).

For the configuration cleanup suggestion, are you saying

1) Move the DAILY_INDEX_FORMAT constant from IndexNameFormatter to ElasticsearchConfig? Not sure if this reduces clutter.

2) Move the zipkin_template.json file into code that is returned in Config? I understand the feeling, but find it pretty impossible to maintain a large JSON string in Java code, so went with resource loading. If you'd still prefer it to be in code, I can do it though.

adriancole commented:
> 1) Move the DAILY_INDEX_FORMAT constant from IndexNameFormatter to
ElasticsearchConfig? Not sure if this reduces clutter.
By clutter I meant a file which only holds constants derived from parts of
config. If I'm not mistaken, IndexNameFormatter only held constants. If you
feel strongly against, feel free to skip this as I don't feel strongly.

> 2) Move the zipkin_template.json file into code that is returned in
Config? I understand the feeling, but find it pretty impossible to maintain
a large JSON string in Java code, so went with resource loading. If you'd
still prefer it to be in code, I can do it though.
I didn't mean to write it in java, rather handle the resource loading of
the string in the config class.

Thinking was that loading a string from a resource isn't critical logic to
understand while reading the implementation. How do you feel about it?

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/IndexNameFormatter.java:
```diff
@@ -0,0 +1,39 @@
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+final class IndexNameFormatter {
```
oops nevermind about this. It isn't actually a constant holder! my bad

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchConfig.java:
```diff
@@ -0,0 +1,65 @@
+    clusterName = checkNotNull(builder.cluster, "builder.cluster");
+    hosts = checkNotNull(builder.hosts, "builder.hosts");
+    index = checkNotNull(builder.index, "builder.index");
+  }
```
I was thinking about having the indexTemplate load here...

```java
try {
this.indexTemplate = Resources.toString(Resources.getResource("zipkin/elasticsearch/zipkin_template.json"), UTF_8);
} catch (IOException e) {
 throw new AssertionError("Error reading jar resource, shouldn't happen.", e);
}
```

adriancole replied:
especially since you know the index already, you can replace it here, too `....replace("${__INDEX__}", index);`

anuraaga replied:
Got it, thanks! I misunderstood in the context of "reducing extra files" but that was talking about the formatter, not this json file.

I've moved the resource read into this Config

adriancole commented on zipkin-spanstores/elasticsearch/src/main/java/zipkin/elasticsearch/ElasticsearchSpanStore.java:
```diff
@@ -0,0 +1,426 @@
+      groupedSpans.put(span.traceId, span);
+    }
+    ImmutableList.Builder<List<Span>> traces = ImmutableList.builder();
+    // We want to return traces in the same order as the parameter.
```
I think this isn't necessary since there's no multi-get anymore

anuraaga replied:
This is a private method shared by both getTraces (which has multiple) and getTrace, the comment refers mostly to when being called from getTraces, though is technically true when there's only a single traceId present.

adriancole commented:
ok I think I've done as much reviewing as possible before merge.

any remaining notes I made, feel free to do or skip. One thing that would be nice is to update the root-level README file to say that elasticsearch is supported.

:clap:

adriancole commented:
once you say merge.. I merge

anuraaga commented:
I've moved the resource load into Config and added the note to the top-level README. I think this is ready to merge.

adriancole commented:
SG merging on green (drum roll begins...)

shakuzen commented on zipkin-server/src/main/java/zipkin/server/ZipkinElasticsearchProperties.java:
```diff
@@ -0,0 +1,64 @@
+   * format. The port should be the transport port, not the http port. Defaults to
+   * "localhost:9300".
+   */
+  private String hosts = "localhost:9300";
```
For some type safety, you could make this a `List<String>` which might simplify things where it is used and catch configuration mistakes sooner. The data binder will convert a comma-separated list to a `List` for you.

anuraaga replied:
Is there a good way to be able to set a ```List<String>``` from an environment variable, as we do here?

https://github.com/openzipkin/zipkin-java/blob/master/zipkin-server/src/main/resources/zipkin-server.yml#L28

If spring-boot would automatically convert a comma-list in that environment variable into List<String> would love to use it!

shakuzen replied:
Sure. That same configuration will work as is. See a demo I have provided [in this gist](https://gist.github.com/shakuzen/da57f99bc480725a5747).

adriancole commented:
those too impatient to wait for a release can play with the latest snapshot exec jar

http://oss.jfrog.org/artifactory/oss-snapshot-local/io/zipkin/java/zipkin-server/0.9.4-SNAPSHOT/

adriancole commented:
on the way out in 0.10.0

normal <1.0 disclaimers apply
